lib/openstack_query/query_wrapper.py

```diff
@@ -0,0 +1,353 @@
+        items: List[Any],
+        property_funcs: Dict[str, Callable[[Any], Any]],
+    ) -> List[Dict[str, str]]:
+        """
+        Generates a dictionary of queried properties from a list of openstack objects e.g. servers
+        :param items: List of items to obtain properties from
+        :param property_funcs: property_functions that return properties requested
+        :return: List containing dictionaries of the requested properties obtained from the items
+        """
+        return [QueryWrapper._parse_property(item, property_funcs) for item in items]
+
+    @staticmethod
+    def _parse_property(
+        item: Any, property_funcs: Dict[str, Callable[[Any], Any]]
+    ) -> Dict[str, str]:
+        """
+        Generates a dictionary of queried properties from a single openstack object
+        :param item: openstack resource item to obtain properties from
+        :param property_funcs: Dict of 'property_name' 'property_function' key value pairs
+        """
+        return {
+            prop_key: QueryWrapper._run_prop_func(
+                item, prop_func, default_out="Not Found"
+            )
+            for prop_key, prop_func in property_funcs.items()
+        }
+
+    @staticmethod
+    def _run_prop_func(
+        item: Any, prop_func: Callable[[Any], Any], default_out="Not Found"
+    ) -> str:
+        """
+        Runs a function to get a property for a given openstack resource
+        :param item: openstack resource item to obtain property from
+        :param prop_func: A 'property_function' to return the property of an openstack resource
+        :param default_out: A default to set if the property cannot be found on the resource
+        """
+        try:
+            return str(prop_func(item))
+        except AttributeError:
+            return default_out
+
+    @staticmethod
+    def _apply_filter_func(
+        items: List[Any], query_func: Callable[[Any], bool]
+    ) -> List[Any]:
+        """
+        Removes items from a list by running a given query function
+        :param items: List of items to query e.g. list of servers
+        :param query_func: Query function that determines whether a given item
+                           matches the query - should return true if it passes
+                           the query
+        :return: List of items that match the given query
+        """
+        return [item for item in items if query_func(item)]
+
+    @staticmethod
+    def _run_query(
+        self,
+        conn: OpenstackConnection,
+        filter_kwargs: Optional[Dict[str, str]] = None,
+        **kwargs,
+    ) -> List[Any]:
+        """
+        This method is to be instantiated in subclasses of this class to run openstack query command
+        """
+        raise NotImplementedError(
+            """
+            This static method should be implemented in subclasses of QueryWrapper to
+            run the appropriate openstack command(s)
+        """
+        )
```

Comment:
Anything related to not building the query, e.g. running it, should be split out of this class into a dedicated runner instead even if this is an abstract method

Also this should be annotated with `@abstractmethod`

Reply:
Split out query wrapper into builder, runner and output - query_wrapper no instantiates these classes and coordinates on requests to them

QueryWrapper now only maintains a set of public methods which we expose to the users as our api